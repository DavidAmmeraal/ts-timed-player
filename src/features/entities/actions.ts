/**
 * actions.ts
 * Action creators for entities actions.
 */
import Types from 'Types';
import { createAction } from 'typesafe-actions';

export const createEntityAction = createAction(
  'entities/CREATE_ENTITY',
  resolve => (entityType: string, props: Types.EntityProps) =>
    resolve({
      entityType,
      props,
    }),
);

export const updateEntityAction = createAction(
  'entities/UPDATE_ENTITY',
  resolve => (entityType: string, props: Types.EntityProps) =>
    resolve({
      entityType,
      props,
    }),
);

export const deleteEntityAction = createAction(
  'entities/DELETE_ENTITY',
  resolve => (entityType: string, id: string) =>
    resolve({
      entityType,
      id,
    }),
);
