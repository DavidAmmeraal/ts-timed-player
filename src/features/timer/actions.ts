/**
 * actions.ts
 * Action creators for timer.
 */
import { action } from 'typesafe-actions';

import { TOGGLE, TICK, SET_RATE, SET_TIME } from './constants';

export const toggle = () => action(TOGGLE);
export const tick = (time: number) => action(TICK, time);
export const setRate = (rate: number) => action(SET_RATE, rate);
/**
 * Creates an SET_TIME time action.
 * @param time time in microseconds
 */
export const setTime = (time: number) => action(SET_TIME, time);
