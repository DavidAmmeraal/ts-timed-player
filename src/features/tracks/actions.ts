import { action } from 'typesafe-actions';

import { CREATE_TRACK, UPDATE_TRACK, DELETE_TRACK } from './constants';
import { ITrack } from './models';

export const createTrack = (track:ITrackEntity) => action(CREATE_TRACK, track);
export const updateTrack = (id:string, track: Partial<ITrack>) => action(UPDATE_TRACK, {
    id,
    track,
});
export const deleteTrack = (id:string) => action(DELETE_TRACK, id);
