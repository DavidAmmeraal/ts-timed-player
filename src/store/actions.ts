import { createAsyncAction } from 'typesafe-actions';

import { FETCH_STAGE_REQUEST, FETCH_STAGE_SUCCESS, FETCH_STAGE_ERROR } from './constants';

export const fetchStage = createAsyncAction(
  FETCH_STAGE_REQUEST,
  FETCH_STAGE_SUCCESS,
  FETCH_STAGE_ERROR,
)<void, {}, Error>();
