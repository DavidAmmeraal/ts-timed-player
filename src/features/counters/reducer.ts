import { ActionType } from 'typesafe-actions';
import * as actions from './actions';
import { ADD, INCREMENT } from './constants';

export type CountersAction = ActionType<typeof actions>;
export type CountersState = {
  readonly value: number;
};

export const initialState: CountersState = {
  value: 1,
};

export default (state: CountersState = initialState, action: CountersAction) => {
  switch (action.type) {
    case ADD:
      return {
        ...state,
        value: state.value + action.payload,
      };
    case INCREMENT:
      return {
        ...state,
        value: state.value + 1,
      };
    default:
      return state;
  }
};
