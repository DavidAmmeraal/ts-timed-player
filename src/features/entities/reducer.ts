/**
 * reducer.ts
 * Handles the entities actions.
 */
import Types from 'Types';
import * as actions from './actions';
import { ActionType } from 'typesafe-actions';
import { EntitiesContainer, createEntitiesContainer } from './models';
import { CREATE, UPDATE, DELETE } from './constants';
import {
  pipe,
  unless,
  has,
  lensProp,
  set,
  lensPath,
  over,
  union,
  when,
  view,
  dissoc,
  without,
} from 'ramda';

export type EntitiesAction = ActionType<typeof actions>;
type CreateAction = ActionType<typeof actions.createEntityAction>;
type UpdateAction = ActionType<typeof actions.updateEntityAction>;
type DeleteAction = ActionType<typeof actions.deleteEntityAction>;

export interface EntitiesState {
  [key: string]: EntitiesContainer;
}
export const defaultState: EntitiesState = {};

/**
 * Handles entities/CREATE actions. Creates EntitiesContainer for entity type if it doesn't yet,
 * and then adds the new entity and its id to the list of ids.
 *
 * @param state Base state.
 * @param action entities/CREATE action to be handled.
 * @returns new state.
 */
const handleCreate = (state: EntitiesState, { payload: { entityType, entity } }: CreateAction) =>
  pipe<EntitiesState, EntitiesState, EntitiesState, EntitiesState>(
    unless(has(entityType), set(lensProp(entityType), createEntitiesContainer())),
    set(lensPath([entityType, 'byId', entity.id]), entity),
    over(lensPath([entityType, 'ids']), union([entity.id])),
  )(state);

/**
 * Handles entities/UPDATE actions. Updates (shallow merge) an entity with given id.
 * Returns input state if entity does not exist.
 *
 * @param state Base state.
 * @param action entities/UPDATE action to be handled.
 * @returns new state.
 */
const handleUpdate = (state: EntitiesState, { payload: { entityType, id, props } }: UpdateAction) =>
  pipe<EntitiesState, EntitiesState>(
    when(
      view(lensPath([entityType, 'byId', id])),
      over(lensPath([entityType, 'byId', id]), existing => ({ ...existing, ...props })),
    ),
  )(state);

/**
 * Handles entities/DELETE actions. Deletes an entity with id.
 *
 * @param state Base state.
 * @param action Entities/DELETE action to be handled.
 */
const handleDelete = (state: EntitiesState, { payload: { entityType, id } }: DeleteAction) =>
  pipe<EntitiesState, EntitiesState, EntitiesState>(
    over(lensPath([entityType, 'byId']), dissoc(id)),
    over(lensPath([entityType, 'ids']), without([id])),
  )(state);

/**
 * Handles and reduces actions into state.
 *
 * @param state The current state.
 * @param action The action to be handled.
 */
export const reducer = (state: EntitiesState = defaultState, action: Types.RootAction) => {
  switch (action.type) {
    case CREATE:
      return handleCreate(state, action);
    case UPDATE:
      return handleUpdate(state, action);
    case DELETE:
      return handleDelete(state, action);
    default:
      return state;
  }
};
