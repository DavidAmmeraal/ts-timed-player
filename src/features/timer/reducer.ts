import * as actions from './actions';
import { ActionType } from 'typesafe-actions';
import { TOGGLE, TICK, SET_RATE, SET_TIME } from './constants';

export type TimerAction = ActionType<typeof actions>;

export type TimerState = {
  readonly time: number;
  readonly rate: number;
  readonly running: boolean;
};

export const defaultState: TimerState = {
  time: 0,
  rate: 1,
  running: false,
};

export default (state: TimerState = defaultState, action: TimerAction) => {
  switch (action.type) {
    case TOGGLE:
      return {
        ...state,
        running: !state.running,
      };
    case TICK:
      return {
        ...state,
        time: state.time + state.rate * action.payload,
      };
    case SET_RATE:
      return {
        ...state,
        rate: action.payload,
      };
    case SET_TIME:
      return {
        ...state,
        time: action.payload,
      };
    default:
      return state;
  }
};
