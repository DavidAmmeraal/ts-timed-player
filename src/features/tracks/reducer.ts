import * as actions from './actions';

export default createEntitiesReducer({
    create: typeof actions.createTrack
})