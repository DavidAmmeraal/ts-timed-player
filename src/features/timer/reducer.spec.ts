import reducer from './reducer';
import * as actions from './actions';
/**
 * FIXTURES
 */
const initialState = reducer(undefined, {} as any);

describe('timer reducer', () => {
    it('should have initial state', () => {
        expect(initialState).toMatchSnapshot();
    })

    it('should return same state on any action not being handled', () => {
        const state = reducer(initialState, {} as any);
        expect(state).toBe(initialState);
    })

    describe('handling TOGGLE action', () => {
        it('should toggle the running variable', () => {
            const action = actions.toggle();
            const state = reducer(initialState, action);
            expect(state.running).toBe(true);
            const toggledState = reducer(state, action);
            expect(toggledState.running).toBe(false);
        })
    })

    describe('handling TICK action', () => {
        it('should increase time by payload', () => {
            const timeA = 1.22;
            const tickA = actions.tick(timeA);

            const timeB = 0.99;
            const tickB = actions.tick(timeB);
            const state = reducer(initialState, tickA);
            const resultState = reducer(state, tickB);
            expect(resultState.time).toBe(timeA + timeB);
        })

        it('should multiply payload by rate', () => {
            const rate = 0.35;
            const time = 0.45

            const state = {
                ...initialState,
                time,
                rate
            };

            const tickTime = 1.22;
            const tick = actions.tick(tickTime);

            const resultState = reducer(state, tick);
            expect(resultState.time).toBe(time + (tickTime * rate));
        })
    });

    describe('handling SET_RATE action', () => {
        it('should set rate to payload', () => {
            const rate = initialState.rate + 3.55;
            const setRateAction = actions.setRate(rate);
            const resultState = reducer(initialState, setRateAction);
            expect(resultState.rate).toBe(rate);
        })
    })

    describe('handling SET_TIME action', () => {
        it('should set time to payload', () => {
            const time = initialState.time + 13.44;
            const setTimeAction = actions.setTime(time);
            const resultState = reducer(initialState, setTimeAction);
            expect(resultState.time).toBe(time);
        })
    });
});