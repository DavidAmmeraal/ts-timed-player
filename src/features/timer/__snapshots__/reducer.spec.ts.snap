// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timer reducer should have initial state 1`] = `
Object {
  "rate": 1,
  "running": false,
  "time": 0,
}
`;
